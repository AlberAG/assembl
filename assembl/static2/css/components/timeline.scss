.timeline {
  height: $navbar-height;

  .bar {
    background-color: $first-color;
    bottom: 0;
    height: 10px;
    position: absolute;
    z-index: 2;
  }

  .bar-bkg {
    background-color: $grey1;
    bottom: 0;
    height: 10px;
    position: absolute;
    width: 100%;
  }

  .pointer {
    fill: $first-color;
    position: absolute;
    z-index: 10;
  }

  .trsp-pointer {
    height: 51px;
  }

  .timeline-date {
    color: $first-color;
    font-size: $font-size-m;
    letter-spacing: $text-letter-spacing;
    margin-top: -30px;
    position: absolute;
    white-space: nowrap;
  }
}

.minimized-timeline {
  height: 100%;
  position: relative;
  flex-grow: 1;
  flex-basis: 0;
  display: flex;
  flex-direction: column-reverse;

  .active {
    background-color: $first-color;
  }

  .not-active {
    background-color: $grey5;
  }

  .timeline-number {
    border: 2px solid $white;
    border-radius: 18px;
    display: inline-block;
    color: $white;
    font-size: $font-size-xxxs;
    height: 18px;
    line-height: 14px;
    padding: 0 0 0 1px;
    position: relative;
    text-align: center;
    width: 18px;
    z-index: 2;

    .assembl-icon-checked {
      &:before {
        font-size: $font-size-xxxs;
      }
    }
  }

  .timeline-graph {
    position: relative;

    .timeline-bars {
      display: inline-block;
      width: calc(100% - 10px);
      height: auto;
      position: absolute;
      top: 9px;
      left: 12px;

      .timeline-bar-background {
        background-color: $grey5;
        height: 6px;
        position: absolute;
        width: 100%;
        z-index: 0;
        display: inline-block;
      }

      .timeline-bar-filler {
        background-color: $first-color;
        height: 6px;
        position: absolute;
        z-index: 1;
        display: inline-block;
      }
    }
  }

  .timeline-title {
    cursor: pointer;
    font-size: 10px;
    padding: 0 0 0 15px;
    text-align: center;
    top: 28px;

    .timeline-link {
      height: 100%;
    }
  }

  .txt-not-active {
    .timeline-link {
      color: $grey5;
    }

    &:hover {
      ~ .timeline-graph {
        .timeline-number {
          background-color: $first-color;
        }

        .timeline-bar-background {
          background-color: $first-color;
        }
      }

      .timeline-link {
        color: $first-color;
      }
    }
  }

  .txt-active-light {
    .timeline-link {
      color: $first-color;
      font-family: $font-family4;

      &:hover {
        font-family: $font-family;
      }
    }
  }

  .txt-active-bold {
    .timeline-link {
      color: $first-color;
      font-family: $font-family3;
    }
  }

  .assembl-icon-down-dir:before {
    font-size: 15px;
  }
}
