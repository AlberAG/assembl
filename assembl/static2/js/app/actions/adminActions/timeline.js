// @flow
import { type moment } from 'moment';
import * as actionTypes from '../actionTypes';

export const createPhase = (id: string, order: number): actionTypes.CreatePhase => ({
  id: id,
  order: order,
  type: actionTypes.CREATE_PHASE
});

export const updatePhases = (phases: actionTypes.PhasesArray): actionTypes.UpdatePhases => ({
  phases: phases,
  type: actionTypes.UPDATE_PHASES
});

export const updatePhaseTitle = (id: string, locale: string, value: string): actionTypes.UpdatePhaseTitle => ({
  id: id,
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_PHASE_TITLE
});

export const updatePhaseIdentifier = (id: string, identifier: string): actionTypes.UpdatePhaseIdentifier => ({
  id: id,
  value: identifier,
  type: actionTypes.UPDATE_PHASE_IDENTIFIER
});


export const deletePhase = (id: string): actionTypes.DeletePhase => ({
  id: id,
  type: actionTypes.DELETE_PHASE
});

export const updateStartDate = (id: string, value: moment): actionTypes.UpdatePhaseStart => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_PHASE_START
});

export const updateEndDate = (id: string, value: moment): actionTypes.UpdatePhaseEnd => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_PHASE_END
});

export const updateIsThematicsTable = (id: string, value: boolean): actionTypes.UpdateIsThematicsTable => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_IS_THEMATICS_TABLE
});

export const movePhaseUp = (id: string): actionTypes.MovePhaseUp => ({
  id: id,
  type: actionTypes.MOVE_PHASE_UP
});

export const movePhaseDown = (id: string): actionTypes.MovePhaseDown => ({
  id: id,
  type: actionTypes.MOVE_PHASE_DOWN
});