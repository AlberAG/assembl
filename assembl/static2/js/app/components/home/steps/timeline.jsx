import React from 'react';
import { connect } from 'react-redux';
import { Localize } from 'react-redux-i18n';
import MapStateToProps from '../../../store/mapStateToProps';
import GlobalFunctions from '../../../utils/globalFunctions';
import Pointer from '../../svg/pointer';

class Timeline extends React.Component {
  getBarWidth() {
    const { debateData } = this.props.debate;
    const index = this.props.index;
    const currentDate = new Date();
    const endDate = GlobalFunctions.getDateFromString(debateData.config.home.steps.timeline[index].endDate);
    const isStepCompleted = GlobalFunctions.isDateExpired(currentDate, endDate);
    let barWidth = 0;
    if (isStepCompleted) {
      barWidth = 100;
    } else {
      const startDate = GlobalFunctions.getDateFromString(debateData.config.home.steps.timeline[index].startDate);
      const isStepStarted = GlobalFunctions.isDateExpired(currentDate, startDate);
      if (isStepStarted) {
        const remainingDays = GlobalFunctions.getNumberOfDays(endDate, currentDate);
        const totalDays = GlobalFunctions.getNumberOfDays(endDate, startDate);
        const percentage = GlobalFunctions.calculatePercentage(remainingDays, totalDays);
        barWidth = 100 - percentage;
      }
    }
    return barWidth;
  }
  render() {
    const currentStep = this.props.currentStep;
    const index = this.props.index;
    const barWidth = this.getBarWidth();
    const currentDate = new Date();
    const datePosition = 100 - barWidth;
    return (
      <div className="timeline">
        {currentStep &&
          <div>
            <div className="timeline-date" style={index === 0 ? { left: `${barWidth}%` } : { right: `${datePosition}%` }}>
              <Localize value={currentDate} dateFormat="date.format" />
            </div>
            <Pointer position={barWidth} />
          </div>
        }
        {!currentStep &&
          <div className="trsp-pointer">&nbsp;</div>
        }
        <div className="bar" style={{ width: `${barWidth}%` }}>&nbsp;</div>
        <div className="bar-bkg">&nbsp;</div>
      </div>
    );
  }
}

export default connect(MapStateToProps)(Timeline);