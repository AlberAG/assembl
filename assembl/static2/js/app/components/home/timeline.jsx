import React from 'react';
import { Grid } from 'react-bootstrap';

class Timeline extends React.Component {
  render() {
    return (
      <Grid fluid>
        <div className="max-container background-grey">
          <div className="timeline" style={{ height: `${200}px` }}>Timeline</div>
        </div>
      </Grid>
    );
  }
}

export default Timeline;