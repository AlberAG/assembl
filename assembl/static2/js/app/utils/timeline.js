import { isDateExpired, getNumberOfDays, calculatePercentage } from './globalFunctions';

export const getCurrentPhaseIdentifier = (_timeline) => {
  let timeline = _timeline;
  if (!timeline) {
    timeline = [];
  }
  const currentDate = new Date();
  let identifier = '';
  timeline.forEach((phase) => {
    const startDate = new Date(phase.start);
    const endDate = new Date(phase.end);
    if (currentDate > startDate && currentDate < endDate) {
      identifier = phase.identifier;
    }
  });
  return identifier || 'thread';
};

export const isPhaseStarted = (_timeline, _identifier) => {
  let timeline = _timeline;
  const identifier = _identifier;
  if (!timeline) {
    timeline = [];
  }
  const currentDate = new Date();
  let isStarted = false;
  timeline.forEach((phase) => {
    if (phase.identifier === identifier) {
      const startDate = new Date(phase.start);
      isStarted = isDateExpired(currentDate, startDate);
    }
  });
  return isStarted;
};

let key = '';
export const isSeveralIdentifiers = (_timeline) => {
  let timeline = _timeline;
  if (!timeline) {
    timeline = [];
  }
  let isSeveral = false;
  timeline.forEach((phase) => {
    if (key !== phase.identifier && key !== '') {
      isSeveral = true;
    }
    key = phase.identifier;
  });
  return isSeveral;
};

export const getPhaseStatus = (_startDate, _endDate) => {
  const startDate = new Date(_startDate);
  const endDate = new Date(_endDate);
  const currentDate = new Date();
  let phaseStatus = '';
  if (currentDate < startDate && currentDate < endDate) {
    phaseStatus = 'notStarted';
  }
  if (currentDate > startDate && currentDate < endDate) {
    phaseStatus = 'inProgress';
  }
  if (currentDate > startDate && currentDate > endDate) {
    phaseStatus = 'completed';
  }
  return phaseStatus;
};

export const getStartDatePhase = (_timeline, _identifier) => {
  let timeline = _timeline;
  if (!timeline) {
    timeline = [];
  }
  const identifier = _identifier;
  let startDatePhase = '';
  timeline.forEach((phase) => {
    if (phase.identifier === identifier) {
      startDatePhase = phase.start;
    }
  });
  return startDatePhase;
};

export const isCurrentPhase = (_phase) => {
  const phase = _phase;
  const currentDate = new Date();
  const startDate = new Date(phase.start);
  const endDate = new Date(phase.end);
  const currentPhase = isDateExpired(currentDate, startDate) && isDateExpired(endDate, currentDate);
  return currentPhase;
};

export const getPhaseName = (_timeline, _identifier, _locale) => {
  let timeline = _timeline;
  if (!timeline) {
    timeline = [];
  }
  const identifier = _identifier;
  const locale = _locale;
  let phaseName = '';
  timeline.forEach((phase) => {
    if (phase.identifier === identifier) {
      phase.title.entries.forEach((entry) => {
        if (entry['@language'] === locale) {
          phaseName = entry.value;
        }
      });
    }
  });
  return phaseName;
};

export const isStepCompleted = (_phase) => {
  const phase = _phase;
  const currentDate = new Date();
  const endDate = new Date(phase.end);
  return isDateExpired(currentDate, endDate);
};

export const getIfPhaseCompletedByIdentifier = (_timeline, _identifier) => {
  let timeline = _timeline;
  if (!timeline) {
    timeline = [];
  }
  const identifier = _identifier;
  let isPhaseCompleted = false;
  timeline.forEach((phase) => {
    if (identifier === phase.identifier) {
      isPhaseCompleted = isStepCompleted(phase);
    }
  });
  return isPhaseCompleted;
};

export const getBarWidth = (_phase) => {
  const phase = _phase;
  const currentDate = new Date();
  const endDate = new Date(phase.end);
  const stepCompleted = isDateExpired(currentDate, endDate);
  let barWidth = 0;
  if (stepCompleted) {
    barWidth = 100;
  } else {
    const startDate = new Date(phase.start);
    const isStepStarted = isDateExpired(currentDate, startDate);
    if (isStepStarted) {
      const remainingDays = getNumberOfDays(endDate, currentDate);
      const totalDays = getNumberOfDays(endDate, startDate);
      const percentage = calculatePercentage(remainingDays, totalDays);
      barWidth = 100 - percentage;
    }
  }
  return barWidth;
};