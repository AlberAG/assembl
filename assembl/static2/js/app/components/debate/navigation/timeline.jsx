// @flow
import React from 'react';
import { connect } from 'react-redux';
import { isCurrentPhase, getBarPercent, isStepCompleted } from '../../../utils/timeline';
import TimelineSegment, { type DebateType } from './timelineSegment';

type TimelineProps = {
  debate: DebateType,
  showNavigation: boolean,
  identifier: string,
  onMenuItemClick: Function
};

export function DumbTimeline(props: TimelineProps) {
  const { debateData } = props.debate;
  const { showNavigation, identifier, onMenuItemClick } = props;
  return (
    <div className="timeline-container">
      {debateData.timeline &&
        debateData.timeline.map((phase, index) => (
          <TimelineSegment
            title={phase.title}
            index={index}
            key={index}
            barPercent={getBarPercent(debateData.timeline[index])}
            isCurrentPhase={isCurrentPhase(debateData.timeline[index])}
            showNavigation={showNavigation}
            identifier={identifier}
            phaseIdentifier={phase.identifier}
            startDate={phase.start}
            endDate={phase.end}
            isStepCompleted={isStepCompleted(phase)}
            onMenuItemClick={onMenuItemClick}
          />
        ))}
    </div>
  );
}

const mapStateToProps = state => ({
  debate: state.debate
});

export default connect(mapStateToProps)(DumbTimeline);