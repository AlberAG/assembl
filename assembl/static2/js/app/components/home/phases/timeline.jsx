import React from 'react';
import { connect } from 'react-redux';
import { Localize } from 'react-redux-i18n';
import { isDateExpired, getNumberOfDays, calculatePercentage } from '../../../utils/globalFunctions';
import Pointer from '../../svg/pointer';

class Timeline extends React.Component {
  getBarWidth() {
    const { debateData } = this.props.debate;
    const index = this.props.index;
    const currentDate = new Date();
    const endDate = new Date(debateData.timeline[index].end);
    const isStepCompleted = isDateExpired(currentDate, endDate);
    let barWidth = 0;
    if (isStepCompleted) {
      barWidth = 100;
    } else {
      const startDate = new Date(debateData.timeline[index].start);
      const isStepStarted = isDateExpired(currentDate, startDate);
      if (isStepStarted) {
        const remainingDays = getNumberOfDays(endDate, currentDate);
        const totalDays = getNumberOfDays(endDate, startDate);
        const percentage = calculatePercentage(remainingDays, totalDays);
        barWidth = 100 - percentage;
      }
    }
    return barWidth;
  }
  render() {
    const currentStep = this.props.currentStep;
    const index = this.props.index;
    const barWidth = this.getBarWidth();
    const currentDate = new Date();
    const datePosition = 100 - barWidth;
    return (
      <div className="timeline">
        {currentStep &&
          <div>
            <div className="timeline-date" style={index === 0 ? { left: `${barWidth}%` } : { right: `${datePosition}%` }}>
              <Localize value={currentDate} dateFormat="date.format" />
            </div>
            <Pointer position={barWidth} />
          </div>
        }
        {!currentStep &&
          <div className="trsp-pointer">&nbsp;</div>
        }
        <div className="bar" style={{ width: `${barWidth}%` }}>&nbsp;</div>
        <div className="bar-bkg">&nbsp;</div>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    debate: state.debate
  };
};

export default connect(mapStateToProps)(Timeline);