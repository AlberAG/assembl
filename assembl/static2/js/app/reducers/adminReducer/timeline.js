// @flow
import type ReduxAction from 'redux';
import { combineReducers } from 'redux';
import moment from 'moment';
import { List, Map, fromJS } from 'immutable';
import { updateInLangstringEntries } from '../../utils/i18n';

import {
  type Action,
  CREATE_PHASE,
  UPDATE_PHASES,
  DELETE_PHASE,
  UPDATE_PHASE_TITLE,
  UPDATE_PHASE_IDENTIFIER,
  UPDATE_PHASE_START,
  UPDATE_PHASE_END,
  UPDATE_IS_THEMATICS_TABLE
} from '../../actions/actionTypes';

const emptyPhase = Map({
  _hasChanged: false,
  _isNew: true,
  _toDelete: false,
  identifier: '',
  titleEntries: List(),
  start: moment(),
  end: moment(),
  isThematicsTable: false
});

type PhasesByIdState = Map<string, Map>;
type PhasesByIdReducer = (PhasesByIdState, ReduxAction<Action>) => PhasesByIdState
export const phasesById: PhasesByIdReducer = (state: PhasesByIdState = Map(), action: ReduxAction<Action>) => {
  switch (action.type) {
  case CREATE_PHASE:
    return state.set(action.id, emptyPhase.set('id', action.id));
  case DELETE_PHASE:
    return state.setIn([action.id, '_toDelete'], true);
  case UPDATE_PHASE_TITLE:
    return state.updateIn([action.id, 'titleEntries'], updateInLangstringEntries(action.locale, action.value));
  case UPDATE_PHASE_IDENTIFIER:
    return state.setIn([action.id, 'identifier'], action.value);
  case UPDATE_PHASE_START:
    return state.setIn([action.id, 'start'], action.value);
  case UPDATE_PHASE_END:
    return state.setIn([action.id, 'end'], action.value);
  case UPDATE_IS_THEMATICS_TABLE:
    return state.setIn([action.id, 'isThematicsTable'], action.value);
  case UPDATE_PHASES: {
    let newState = Map();
    action.phases.forEach(({ identifier, titleEntries, start, end, id, isThematicsTable }) => {
      const phaseInfo = Map({
        _hasChanged: false,
        _isNew: false,
        _toDelete: false,
        identifier: identifier,
        titleEntries: fromJS(titleEntries),
        start: start,
        end: end,
        isThematicsTable: isThematicsTable
      });

      newState = newState.set(id, phaseInfo);
    });

    return newState;
  }
  default:
    return state;
  }
};

type PhasesInOrderState = List<string>;
type PhasesInOrderReducer = (PhasesInOrderState, ReduxAction<Action>) => PhasesInOrderState;
export const phasesInOrder: PhasesInOrderReducer = (state = List(), action) => {
  switch (action.type) {
  case CREATE_PHASE:
    return state.push(action.id);
  case UPDATE_PHASES:
    return List(action.phases.map(phase => phase.id));
  default:
    return state;
  }
};

type TimelineHasChangedReducer = (boolean, ReduxAction<Action>) => boolean;
export const phasesHaveChanged: TimelineHasChangedReducer = (state = false, action) => {
  switch (action.type) {
  case UPDATE_PHASES:
    return false;
  case CREATE_PHASE:
  case DELETE_PHASE:
  case UPDATE_PHASE_TITLE:
  case UPDATE_PHASE_IDENTIFIER:
  case UPDATE_PHASE_START:
  case UPDATE_PHASE_END:
    return true;
  default:
    return state;
  }
};

export default combineReducers({
  phasesInOrder: phasesInOrder,
  phasesHaveChanged: phasesHaveChanged,
  phasesById: phasesById
});