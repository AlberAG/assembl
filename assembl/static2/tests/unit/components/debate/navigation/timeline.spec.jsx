import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { DumbTimeline } from '../../../../../js/app/components/debate/navigation/timeline';

describe('DumbTimeline component', () => {
  it('should match the DumbTimeline', () => {
    const timeline = [
      {
        identifier: 'foo',
        start: 'date1',
        end: 'date2',
        title: { entries: [{ en: 'Foo' }] }
      },
      {
        identifier: 'bar',
        start: 'date1Bar',
        end: 'date2Bar',
        title: { entries: [{ en: 'Bar' }] }
      }
    ];
    const props = {
      identifier: 'survey',
      showNavigation: true,
      debate: {
        debateData: {
          slug: 'slug',
          timeline: timeline
        }
      }
    };
    const renderer = new ShallowRenderer();
    renderer.render(<DumbTimeline {...props} />);
    const rendered = renderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });
});