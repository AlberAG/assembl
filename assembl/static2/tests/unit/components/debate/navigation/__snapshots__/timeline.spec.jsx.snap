// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DumbTimeline component should match the DumbTimeline 1`] = `
<div
  className="timeline-container"
/>
`;
